File: app/src/main/java/com/example/data/remote/GrokModels.kt
package com.example.data.remote

/**
 * Central registry of xAI Grok model IDs and capability tiers — the single source of truth so the
 * rest of the app never hardcodes a model string. xAI retired the grok-3 / grok-4 generation on
 * 2026-05-15 (old IDs still resolve but are billed at — and behave like — their successors), so the
 * defaults below target the current `grok-4.3` flagship. Swap a constant here to re-tune the whole
 * stack; nothing else needs to change.
 *
 * Source: https://docs.x.ai/docs/models (verified 2026-06-18).
 */
object GrokModels {
    /**
     * Flagship general model: 1M-token context, native vision (image input), reasoning, function
     * calling and structured outputs. Cost/latency is tuned per-call with [GrokReasoning], so this
     * one model covers what grok-4, grok-4-fast and grok-3-mini used to span.
     */
    const val FLAGSHIP = "grok-4.3"

    /** Lightweight coding-tuned model (256k ctx), successor to grok-code-fast-1. */
    const val CODING = "grok-build-0.1"

    /** Image generation models served from `/v1/images/generations`. */
    const val IMAGE = "grok-imagine-image"               // $0.02/image
    const val IMAGE_QUALITY = "grok-imagine-image-quality" // $0.05/image, higher fidelity

    /** Task-specific defaults (all map onto the flagship today; centralised for easy retuning). */
    const val ANALYSIS = FLAGSHIP        // structured bookmark analysis (JSON-schema enforced)
    const val DEEP_ANALYSIS = FLAGSHIP   // long-form contribution/significance/caveats
    const val CHAT = FLAGSHIP            // RAG + Live-Search research assistant
    const val VISION = FLAGSHIP          // image understanding (OCR-adjacent, figure reading)

    /**
     * Text-embedding model. NOTE: as of 2026-06 xAI exposes embeddings only internally (behind the
     * Collections feature) and ships NO public `/v1/embeddings` REST endpoint, so this call is
     * best-effort and will typically return null — Curio's [com.example.data.embedding.EmbeddingProviderSelector]
     * already degrades to on-device EmbeddingGemma / recency fallback when that happens.
     */
    const val EMBEDDING = "grok-embedding-small"

    /** Legacy IDs kept for reference; these auto-redirect server-side. Do not use for new calls. */
    object Legacy {
        const val GROK_3 = "grok-3"
        const val GROK_3_MINI = "grok-3-mini"
        const val GROK_4 = "grok-4"
        const val GROK_2_IMAGE = "grok-2-image"
    }
}

/** Allowed values for the `reasoning_effort` request field on reasoning-capable Grok models. */
object GrokReasoning {
    const val NONE = "none"     // skip thinking tokens entirely — cheapest/fastest
    const val LOW = "low"       // default
    const val MEDIUM = "medium"
    const val HIGH = "high"     // deepest reasoning, highest cost
}

/** Allowed values for [XAiSearchParameters.mode] (Live Search grounding). */
object GrokSearchMode {
    const val OFF = "off"   // never search
    const val AUTO = "auto" // let the model decide
    const val ON = "on"     // always search
}

File: ios/Curio/Curio/Networking/GrokModels.swift
import Foundation

/// Central registry of xAI Grok model IDs and capability tiers — the single source of truth so the
/// rest of the app never hardcodes a model string. Direct port of `object GrokModels` in
/// `data/remote/GrokModels.kt`.
///
/// xAI retired the grok-3 / grok-4 generation on 2026-05-15 (old IDs still resolve but are billed
/// at — and behave like — their successors), so the defaults below target the current `grok-4.3`
/// flagship. Swap a constant here to re-tune the whole stack; nothing else needs to change.
///
/// Caseless namespace enum (CONVENTIONS §1 "Enums as namespaces").
enum GrokModels {
    /// Flagship general model: 1M-token context, native vision (image input), reasoning, function
    /// calling and structured outputs. Cost/latency is tuned per-call with ``GrokReasoning``.
    static let flagship = "grok-4.3"

    /// Lightweight coding-tuned model (256k ctx), successor to grok-code-fast-1.
    static let coding = "grok-build-0.1"

    /// Image generation model served from `/v1/images/generations`. ($0.02/image)
    static let image = "grok-imagine-image"
    /// Higher-fidelity image generation model. ($0.05/image)
    static let imageQuality = "grok-imagine-image-quality"

    /// Structured bookmark analysis (JSON-schema enforced).
    static let analysis = flagship
    /// Long-form contribution/significance/caveats.
    static let deepAnalysis = flagship
    /// RAG + Live-Search research assistant.
    static let chat = flagship
    /// Image understanding (OCR-adjacent, figure reading).
    static let vision = flagship

    /// Text-embedding model. NOTE: as of 2026-06 xAI exposes embeddings only internally (behind the
    /// Collections feature) and ships NO public `/v1/embeddings` REST endpoint, so this call is
    /// best-effort and will typically return `nil` — the embedding provider selector already
    /// degrades to on-device EmbeddingGemma / recency fallback when that happens.
    static let embedding = "grok-embedding-small"
}

/// Allowed values for the `reasoning_effort` request field on reasoning-capable Grok models.
/// Ports `object GrokReasoning`.
enum GrokReasoning {
    static let none = "none"     // skip thinking tokens entirely — cheapest/fastest
    static let low = "low"       // default
    static let medium = "medium"
    static let high = "high"     // deepest reasoning, highest cost
}

/// Allowed values for `XAiSearchParameters.mode` (Live Search grounding). Ports `object
/// GrokSearchMode`.
enum GrokSearchMode {
    static let off = "off"   // never search
    static let auto = "auto" // let the model decide
    static let on = "on"     // always search
}
